File: src/lib/api/outfits.ts
/**
 * Outfit API functions.
 */

import type {
  OutfitCreate,
  OutfitOut,
  OutfitUpdate,
  PaginatedResponse,
} from "./types";

type Fetcher = <T>(path: string, init?: RequestInit) => Promise<T>;

export async function listOutfits(
  params: { q?: string; character_id?: number; page?: number; page_size?: number },
  fetcher: Fetcher,
): Promise<PaginatedResponse<OutfitOut>> {
  const sp = new URLSearchParams();
  if (params.q) sp.set("q", params.q);
  if (params.character_id) sp.set("character_id", String(params.character_id));
  if (params.page) sp.set("page", String(params.page));
  if (params.page_size) sp.set("page_size", String(params.page_size));
  const qs = sp.toString();
  return fetcher(`/api/v1/outfits${qs ? `?${qs}` : ""}`);
}

export async function getOutfit(
  id: number,
  fetcher: Fetcher,
): Promise<OutfitOut> {
  return fetcher(`/api/v1/outfits/${id}`);
}

export async function createOutfit(
  data: OutfitCreate,
  fetcher: Fetcher,
): Promise<OutfitOut> {
  return fetcher("/api/v1/outfits", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function updateOutfit(
  id: number,
  data: OutfitUpdate,
  fetcher: Fetcher,
): Promise<OutfitOut> {
  return fetcher(`/api/v1/outfits/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
}

export async function deleteOutfit(
  id: number,
  fetcher: Fetcher,
): Promise<void> {
  await fetcher(`/api/v1/outfits/${id}`, { method: "DELETE" });
}

File: backend/app/routers/outfits.py
"""Outfit API routes."""

from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session

from app.config import settings
from app.dependencies import get_db
from app.schemas.common import DeleteOrphansResponse, PaginatedResponse
from app.schemas.entities import OutfitCreate, OutfitOut, OutfitUpdate
from app.services import entity_service

router = APIRouter(prefix="/api/v1/outfits", tags=["outfits"])


@router.get("", response_model=PaginatedResponse[OutfitOut])
def list_outfits(
    q: str | None = None, character_id: int | None = None,
    page: int = 1, page_size: int | None = None,
    db: Session = Depends(get_db),
):
    ps = min(page_size or settings.default_page_size, settings.max_page_size)
    items, total = entity_service.list_outfits(db, q=q, character_id=character_id, page=page, page_size=ps)
    return PaginatedResponse(items=items, total=total, page=page, page_size=ps)


@router.post("/delete-orphans", response_model=DeleteOrphansResponse)
def delete_orphan_outfits(db: Session = Depends(get_db)):
    """Delete every Outfit with no Pack association."""
    deleted = entity_service.delete_orphan_outfits(db)
    return DeleteOrphansResponse(deleted=deleted)


@router.get("/{outfit_id}", response_model=OutfitOut)
def get_outfit(outfit_id: int, db: Session = Depends(get_db)):
    result = entity_service.get_outfit(db, outfit_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Outfit not found")
    return result


@router.post("", response_model=OutfitOut, status_code=201)
def create_outfit(body: OutfitCreate, db: Session = Depends(get_db)):
    outfit = entity_service.create_outfit(db, name=body.name, character_id=body.character_id)
    return entity_service.get_outfit(db, outfit.id)


@router.patch("/{outfit_id}", response_model=OutfitOut)
def update_outfit(outfit_id: int, body: OutfitUpdate, db: Session = Depends(get_db)):
    outfit = entity_service.update_outfit(
        db,
        outfit_id,
        name=body.name,
        character_id=body.character_id,
    )
    if outfit is None:
        raise HTTPException(status_code=404, detail="Outfit not found")
    return entity_service.get_outfit(db, outfit.id)


@router.delete("/{outfit_id}", status_code=204)
def delete_outfit(outfit_id: int, db: Session = Depends(get_db)):
    result = entity_service.delete_outfit(db, outfit_id)
    if result == "not_found":
        raise HTTPException(status_code=404, detail="Outfit not found")
